core/ledger/header.go

```diff
@@ -17,9 +17,15 @@ import (
 type WinningHashType byte
 
 const (
-	GenesisHash      WinningHashType = 0
-	WinningTxnHash   WinningHashType = 1
-	WinningBlockHash WinningHashType = 2
+	// The proof of Block proposer validity should exists in previous Block header.
+	// GenesisHash means next Block proposer is GenesisBlockProposer.
+	GenesisHash WinningHashType = 0
+	// WinningTxnHash means next Block proposer is a node on signature chain.
+	WinningTxnHash WinningHashType = 1
+	// WinningNilHash means next Block proposer is the signer of last Block.
+	WinningNilHash WinningHashType = 2
+	// WinningBlockHash means next Block proposer is signer of historical Block.
+	WinningBlockHash WinningHashType = 3
```

Comment:
Is there any reason why we want to separate WinningNilHash from WinningBlockHash?

Reply:
When no signature chain in system, the last block singer will produce and sign next block. WinningNilHash is only used for verifying block singer, the singer must be signer of last block when no signature chain.

When block singer(could be last block signer or consensused by signature chain) is not available, we do block proposer changing according to timeout. WinningBlockHash is only used for tracking the block proposer which is chosen by timeout from historical block.